### PR TITLE
swarm/storage/feed: add context handling/cancellation to Swarm Feeds lookup, fix bad hint lookup bug

### PR DESCRIPTION
**Pull request by jpeletier**
### Summary: 
* This PR adds context handling/cancellation to Swarm's Feed lookup algorithm.
* Fix Feeds lookup bug corner case

          With ❤ from [Epic Labs](https://www.epiclabs.io)
#### Context cancellation and forwarding
 The Lookup function now admits a `context`. This allows for the algorithm process to be stopped by the caller, or otherwise given a deadline/timeout.

#### Lookup bug fix
Fix a lookup bug whereupon a bad hint placed next to a good update would fool the lookup algorithm having it potentially fail in finding the last update. This is a very infrequent case and probably has not been experienced, but could happen. We detected this while reviewing the algorithm code.

A test was added to cover this case.

### CODE REVIEW

File: swarm/storage/feed/lookup/lookup_test.go [hunk 1]
```diff
@@ -195,6 +196,106 @@ func TestBadHint(t *testing.T) {
...
+	// text context cancellation during hint lookup:
+	ctx, cancel = context.WithCancel(context.Background())
+	errc = make(chan error)
+	someHint := lookup.Epoch{
+		Level: 25,
+		Time:  300,
+	}
+
+	readFunc = func(ctx context.Context, epoch lookup.Epoch, now uint64) (interface{}, error) {
+		if epoch == someHint {
```
Comment (nolash) [inline swarm/storage/feed/lookup/lookup_test.go]: Why do you need this check?

File: swarm/storage/feed/lookup/lookup_test.go [hunk 2]
```diff
@@ -195,6 +196,106 @@ func TestBadHint(t *testing.T) {
...
+		errc <- err
+	}()
+
+	cancel()
+
+	if err := <-errc; err != context.Canceled {
+		t.Fatalf("Expected lookup to return a context Cancelled error, got %v", err)
+	}
+
+	// text context cancellation during hint lookup:
```
Comment (nolash) [inline swarm/storage/feed/lookup/lookup_test.go]: Why do you need to test hint lookup explicitly?

File: swarm/storage/feed/lookup/lookup_test.go (hunk 1 above)
Comment (jpeletier) [reply swarm/storage/feed/lookup/lookup_test.go]: This is to trigger a cancellation only when the hint is reached and looked up.

File: swarm/storage/feed/lookup/lookup_test.go (hunk 2 above)
Comment (jpeletier) [reply swarm/storage/feed/lookup/lookup_test.go]: Algorithms check the hint as a special case, thus this test checks that a context can be cancelled also in this case

File: swarm/storage/feed/lookup/lookup_test.go (hunk 1 above)
Comment (nolash) [reply swarm/storage/feed/lookup/lookup_test.go]: I mean, why would `readFunc` in this test be called with something other than the lookup with `someHint`?

Comment (jpeletier) [reply swarm/storage/feed/lookup/lookup_test.go]: The algorithm will call `readFunc()` many times tentatively with different values for `epoch` that at some point will end up reaching `someHint`. It is the call in which `epoch==someHint` the one that is special and want to cancel to test that particular case.

Comment (nolash) [reply swarm/storage/feed/lookup/lookup_test.go]: So the hint isn't tried first as the slot to look for the possible next update?

Comment (jpeletier) [reply swarm/storage/feed/lookup/lookup_test.go]: No, the hint is not tried first.

Comment (nolash) [reply swarm/storage/feed/lookup/lookup_test.go]: cleared up in orange lounge discussion

**Pull request merged**